kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KtTypes.kt

```diff
@@ -0,0 +1,239 @@
+        }
+        parameter.declaresDefaultValue -> {
+          arguments += defaultPrimitiveValue(parameter.rawType)
+          mask = mask or (1 shl (index % Integer.SIZE))
+        }
+        else -> {
+          throw IllegalArgumentException(
+            "No argument provided for a required parameter: $parameter"
+          )
+        }
+      }
+
+      index++
+    }
+
+    if (!isDefault) {
+      @Suppress("UNCHECKED_CAST")
+      return jvm.newInstance(*arguments.toTypedArray()) as R
+    }
+
+    masks += mask
+    arguments.addAll(masks)
+
+    // DefaultConstructorMarker
+    arguments.add(null)
+
+    @Suppress("UNCHECKED_CAST")
+    return jvm.newInstance(*arguments.toTypedArray()) as R
+  }
+
+  companion object {
+    fun create(rawType: Class<*>, kmClass: KmClass): KtConstructor? {
+      val kmConstructor = kmClass.constructors.find { Flag.Constructor.IS_PRIMARY(it.flags) }
+        ?: return null
+      val kmConstructorSignature = kmConstructor.signature?.asString() ?: return null
+      val constructorsBySignature = rawType.declaredConstructors.associateBy { it.jvmMethodSignature }
+      val jvmConstructor = constructorsBySignature[kmConstructorSignature] ?: return null
+      val parameterAnnotations = jvmConstructor.parameterAnnotations
+      val parameterTypes = jvmConstructor.parameterTypes
+      val parameters = kmConstructor.valueParameters.withIndex()
+        .map { (index, kmParam) ->
+          KtParameter(kmParam, index, parameterTypes[index], parameterAnnotations[index].toList())
+        }
+
+      val anyOptional = parameters.any { it.declaresDefaultValue }
+      val actualConstructor = if (anyOptional) {
+        val prefix = jvmConstructor.jvmMethodSignature.removeSuffix(")V")
+        val parameterCount = jvmConstructor.parameterTypes.size
+        val maskParamsToAdd = if (parameterCount == 0) {
+          0
+        } else {
+          (parameterCount + 31) / 32
+        }
+        val defaultConstructorSignature = buildString {
+          append(prefix)
+          repeat(maskParamsToAdd) {
+            append("I")
+          }
+          append(Util.DEFAULT_CONSTRUCTOR_MARKER!!.descriptor)
+          append(")V")
+        }
+        constructorsBySignature[defaultConstructorSignature] ?: return null
+      } else {
+        jvmConstructor
+      }
+
+      return KtConstructor(
+        rawType,
+        kmConstructor,
+        actualConstructor,
+        parameters,
+        anyOptional
+      )
+    }
+  }
+}
+
+internal data class KtProperty(
+  val km: KmProperty,
+  val jvmField: Field?,
+  val jvmGetter: Method?,
+  val jvmSetter: Method?,
+  val jvmAnnotationsMethod: Method?,
+  val parameter: KtParameter?
+) {
+  init {
+    jvmField?.isAccessible = true
+    jvmGetter?.isAccessible = true
+    jvmSetter?.isAccessible = true
+  }
+
+  val name get() = km.name
+
+  val javaType = jvmField?.genericType
+    ?: jvmGetter?.genericReturnType
+    ?: jvmSetter?.genericReturnType
+    ?: error(
```

Comment:
should this be a condition on init that the property has at least one of these non-null?

Reply:
not sure I follow?

Reply:
Umm I think this happens where it should. Disregard me!